packages/puppeteer-base/src/sudo.rs

```diff
@@ -94,6 +96,62 @@ where
         Ok(deposits)
     }
 
+    pub fn sudo_delegations_and_balance_kv_query_result(
+        &self,
+        deps: DepsMut<NeutronQuery>,
+        env: Env,
+        query_id: u64,
+        version: &str,
+    ) -> NeutronResult<Response<NeutronMsg>> {
+        let chunks_len = self
+            .delegations_and_balances_query_id_chunk
+            .keys(deps.storage, None, None, Order::Ascending)
+            .collect::<Vec<_>>()
+            .len();
+        let chunk_id = self
+            .delegations_and_balances_query_id_chunk
+            .load(deps.storage, query_id)?;
+        let (remote_height, kv_results) = {
+            let registered_query_result = get_raw_interchain_query_result(deps.as_ref(), query_id)?;
+            (
+                registered_query_result.result.height,
+                registered_query_result.result.kv_results,
+            )
+        };
+        let data: BalancesAndDelegations = PuppeteerReconstruct::reconstruct(&kv_results, version)?;
+        let new_state = match self
+            .delegations_and_balances
+            .may_load(deps.storage, &env.block.height)?
+        {
+            Some(mut state) => {
+                state
+                    .data
+                    .delegations
+                    .delegations
+                    .extend(data.delegations.delegations);
+                // insert chunk_id to collected_chunks and uniq
+                state.collected_chunks.push(chunk_id);
+                state.collected_chunks.sort();
+                state.collected_chunks.dedup();
+                state
+            }
+            None => BalancesAndDelegationsState {
+                data,
+                remote_height,
+                local_height: env.block.height,
+                timestamp: env.block.time,
+                collected_chunks: vec![chunk_id],
+            },
+        };
+        if new_state.collected_chunks.len() == chunks_len {
+            self.last_complete_delegations_and_balances_key
+                .save(deps.storage, &env.block.height)?;
+        }
```

Comment:
Before overriding last complete key, please check that the key you save is actually newer than previous one.

Reply:
can you imagine `env.block.height` is less than for the any previous tx?

Reply:
uhhh, don't you think we should group chunks by remote height and not by local height?

Reply:
ok, valid 👍